### PR TITLE
Port `log_softmax` to structured kernel

### PR DESCRIPTION
**Pull request by SplitInfinity**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62513
* #62452
* #62372
* __->__ #57374

Differential Revision: [D30240243](https://our.internmc.facebook.com/intern/diff/D30240243)

### CODE REVIEW

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 1]
```diff
@@ -899,8 +899,13 @@ Tensor host_softmax_backward(const Tensor &grad_, const Tensor &output_, int64_t
...
 
-Tensor log_softmax_cuda(const Tensor &input, const int64_t dim, const bool half_to_float){
-  return host_softmax<LogSoftMaxForwardEpilogue,true>(input, dim, half_to_float);
+TORCH_IMPL_FUNC(log_softmax_cuda_out) (
+  const Tensor &input,
+  const int64_t dim,
+  const bool half_to_float
+  const Tensor &output) {
+  auto res = host_softmax<LogSoftMaxForwardEpilogue,true>(input, dim, half_to_float);
+  output.copy_(res);
```
Comment (SplitInfinity) [inline aten/src/ATen/native/cuda/SoftMax.cu:908]: So ideally I would refactor `host_softmax` so that it takes an `out` parameter but that function is also used by `softmax_cuda`, which is not structured. It can be, and the structure (no pun intended) of that kernel is very similar to that of `log_softmax`, so I think the port would be similar. However, it is not in any of the lists mentioned in #55070.

There are two options:
1. Port `_softmax` to structured so that this `host_softmax` function can be refactored to require an `out` parameter.
2. Do not port `_softmax` to structured. Add an optional `out` parameter to `host_softmax` and use that as the output `Tensor` if one is provided.

I prefer number 1. Let me know your thoughts.

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 2]
```diff
@@ -925,6 +930,5 @@ Tensor softmax_backward_cuda(const Tensor &grad, const Tensor &output, int64_t d
   Tensor tmp = grad * output;
   return host_softmax_backward<SoftMaxBackwardEpilogue,false>(tmp, output, dim, half_to_float);
 }
-
```
Comment (SplitInfinity) [inline aten/src/ATen/native/cuda/SoftMax.cu:928]: Oops, I should delete this.

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/cuda/SoftMax.cu:908]: If not in a rush better to do the cluster of ports necessarily to put us in the better end state.

Comment (SplitInfinity) [reply aten/src/ATen/native/cuda/SoftMax.cu:908]: I agree, but why is `_softmax` not on any of the lists in the first place?

Comment (ezyang) [reply aten/src/ATen/native/cuda/SoftMax.cu:908]: It's possible the end users are only reporting the top level functions, and not how they are implemented internally

File: aten/src/ATen/native/SoftMax.cpp [hunk 3]
```diff
@@ -7,7 +7,35 @@
...
+namespace meta {
+TORCH_META_FUNC(_log_softmax) (
+  const Tensor& input,
+  const int64_t dim,
+  const bool half_to_float) {
+  TORCH_CHECK(
+      !half_to_float,
+      "softmax with half to float conversion is not supported on CPU");
+
+  auto input_ = input.contiguous();
```
Comment (ezyang) [inline aten/src/ATen/native/SoftMax.cpp:22]: improper use of contiguous in meta function

File: aten/src/ATen/native/SoftMax.cpp [hunk 4]
```diff
@@ -117,62 +161,62 @@ void host_softmax_backward(
...
-  int64_t dim = maybe_wrap_dim(dim_, input.dim());
 
   if (input.numel() == 0) {
-    return output;
+    return;
   }
-  if (input.dim() == 0)
-    input = input.view(1);
+
+  auto input_ = input.contiguous();
```
Comment (ezyang) [inline aten/src/ATen/native/SoftMax.cpp:175]: nb: if we ever wanna penny pinch refcounts here, use MaybeOwned, e.g., as in https://github.com/pytorch/pytorch/pull/56115/files this does not block merging this PR

Comment (ezyang) [reply aten/src/ATen/native/SoftMax.cpp:175]: a minor style point; personally, I like underscoring the input parameter and then having it un-underscored in the body function; this is mostly based on which variable I'll be referring to a lot and giving it the "better" name. We're very uneven on this everywhere. (also does not block merging the PR)

**Review by ezyang — APPROVED** [review]

**Comment by SplitInfinity** [comment]
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30240243).
